### ISSUE TITLE
Error loading quentinll/lewm-pusht checkpoint: conversion script fails and Google Drive checkpoint link is 404

### ISSUE THREAD

**Issue by Nishant503** [lucas-maes/le-wm#95]
# Error loading `quentinll/lewm-pusht` checkpoint: conversion script fails and Google Drive checkpoint link is 404

Hi, thank you for releasing LeWM!

I'm trying to reproduce the PushT evaluation, but I've run into what appears to be a checkpoint/version mismatch.

## Environment

* Python 3.10.20
* stable-worldmodel 0.0.6
* stable-pretraining 0.1.8
* Installed exactly as described in the README:

```bash
uv venv --python=3.10
source .venv/bin/activate
uv pip install stable-worldmodel[train,env]
```

## Repository

Using the latest `main` branch of this repository.

## Problem 1: Google Drive checkpoint

The README mentions downloading pretrained checkpoints from Google Drive.

However, the provided Drive link currently returns **404 Not Found**, so I cannot obtain `lewm_object.ckpt`.

## Problem 2: Hugging Face checkpoint conversion

I downloaded:

```
quentinll/lewm-pusht
```

which contains:

```
config.json
weights.pt
```

I then tried to run the conversion script from the README.

The first issue was that the Hydra configuration contains `_target_` fields, so I removed those as suggested in other discussions.

The model can then be constructed successfully, but loading the weights fails because the checkpoint parameter names do not match the current model.

For example:

Expected keys:

```
encoder.layers.0.attention.q_proj.weight
...
```

Checkpoint keys:

```
encoder.encoder.layer.0.attention.attention.query.weight


This results in hundreds of missing and unexpected keys during:

```python
model.load_state_dict(state_dict, strict=True)
```

It looks like the Hugging Face checkpoint was generated using a different version (or commit) of `stable-worldmodel` than the one currently installed from PyPI.

## Questions

Could you please clarify:

1. Which exact versions (or commit hashes) of:

   * stable-worldmodel
   * stable-pretraining

   were used to generate the released `weights.pt`?

2. Is there an updated conversion script compatible with the current packages?

3. Has the Google Drive link for `lewm_object.ckpt` moved? If so, could you share the updated link?

4. Is there currently a recommended way to evaluate the pretrained PushT model without manually converting the Hugging Face checkpoint?

Thank you very much for your help!

**Comment by BogdanRajkov** [comment]
Hi, I had the same issue and threw Claude at it. Here's the summary it generated:

--------------------------

The key mismatch is real, but the cause isn't `stable-worldmodel` or `stable-pretraining`. It's `transformers`.

## What's actually going on

`transformers` renamed the ViT submodules in the 5.x line. The released `weights.pt` was saved with the 4.x names, so the encoder keys no longer line up with the model that `spt.backbone.utils.vit_hf` builds today:

| released `weights.pt` (transformers 4.x) | current model (transformers 5.x) |
|---|---|
| `encoder.encoder.layer.N.attention.attention.query.*` | `encoder.layers.N.attention.q_proj.*` |
| `encoder.encoder.layer.N.attention.attention.key.*` | `encoder.layers.N.attention.k_proj.*` |
| `encoder.encoder.layer.N.attention.attention.value.*` | `encoder.layers.N.attention.v_proj.*` |
| `encoder.encoder.layer.N.attention.output.dense.*` | `encoder.layers.N.attention.o_proj.*` |
| `encoder.encoder.layer.N.intermediate.dense.*` | `encoder.layers.N.mlp.fc1.*` |
| `encoder.encoder.layer.N.output.dense.*` | `encoder.layers.N.mlp.fc2.*` |

`embeddings.*`, `layernorm.*`, and everything outside the encoder (`predictor`, `action_encoder`, `projector`, `pred_proj`) are unchanged — which matches what you saw: only the encoder keys are missing/unexpected.

Claude checked this against the published `quentinll/lewm-pusht` weights: with `transformers==4.57.6` they load into the current model with `strict=True` and no remapping at all; with `transformers==5.15.0` you get exactly the mismatch you pasted. So to your question 1 — the `stable-worldmodel` / `stable-pretraining` versions aren't what matters here, `spt 0.1.8` (what you have) is fine. Any env with `transformers<5` loads the released weights as-is.

## Two other things to fix on your side

1. **Upgrade `stable-worldmodel`.** `main`'s `eval.py` loads the policy through `swm.wm.utils.load_pretrained`, which doesn't exist in 0.0.6 — that release has no `stable_worldmodel/wm/utils.py` at all. Please use 0.1.1.
2. **The README's "Loading a checkpoint" section is outdated** . The `_object.ckpt` + `AutoCostModel` flow has been replaced: a checkpoint is now a folder under `$STABLEWM_HOME/checkpoints/<run_name>/` containing `weights.pt` (a plain `state_dict`) and `config.json`, and `load_pretrained` rebuilds the model with Hydra from that `config.json`. So `_target_` has to *stay* in the saved config — it's only the manual constructor calls in the conversion snippet that must drop it. We'll update the README.

## Updated conversion script

This handles both of the above: it retargets the published config at this repo's `jepa.py` / `module.py`, remaps the encoder keys only when the installed `transformers` needs it (so it works on both 4.x and 5.x), and writes the checkpoint in the layout `eval.py` expects.

```bash
#!/usr/bin/env bash
# Convert a pretrained LeWM checkpoint from the HuggingFace mirror into a local
# checkpoint that eval.py can load:
#   ./get_model.sh                                   # -> policy=pusht/lewm
#   ./get_model.sh quentinll/lewm-cube cube/lewm     # -> policy=cube/lewm
set -euo pipefail

HF_REPO="${1:-quentinll/lewm-pusht}"
RUN_NAME="${2:-pusht/lewm}"

# download weights.pt + config.json under $STABLEWM_HOME
HF_SRC="$(python -c 'import sys, stable_worldmodel as swm
from pathlib import Path
print(Path(swm.data.utils.get_cache_dir(), f"hf_{sys.argv[1]}"))' "${HF_REPO##*/}")"
hf download "$HF_REPO" --local-dir "$HF_SRC"

# convert to $STABLEWM_HOME/checkpoints/$RUN_NAME/{weights.pt,config.json}
HF_SRC="$HF_SRC" RUN_NAME="$RUN_NAME" python - <<'PY'
import json, os, torch, stable_pretraining as spt
from omegaconf import OmegaConf
from pathlib import Path
from jepa import JEPA
from module import ARPredictor, Embedder, MLP
import stable_worldmodel as swm

src = Path(os.environ["HF_SRC"])
run_name = os.environ["RUN_NAME"]

# the published config instantiates the stable_worldmodel implementation, retarget
# it to this repo's modules so that eval.py rebuilds the local JEPA
targets = {
    "": "jepa.JEPA",
    "predictor": "module.ARPredictor",
    "action_encoder": "module.Embedder",
    "projector": "module.MLP",
    "pred_proj": "module.MLP",
}
cfg = json.loads((src / "config.json").read_text())
for key, target in targets.items():
    (cfg if key == "" else cfg[key])["_target_"] = target

# hydra keys (_target_, _partial_) are not constructor arguments
kwargs = lambda k: {a: v for a, v in cfg[k].items() if not a.startswith("_")}
mlp = lambda k: MLP(**{**kwargs(k), "norm_fn": torch.nn.BatchNorm1d})
model = JEPA(
    encoder=spt.backbone.utils.vit_hf(**kwargs("encoder")),
    predictor=ARPredictor(**kwargs("predictor")),
    action_encoder=Embedder(**kwargs("action_encoder")),
    projector=mlp("projector"),
    pred_proj=mlp("pred_proj"),
)

# transformers>=5 renamed the ViT submodules, the published weights use the old names
vit_renames = [
    ("encoder.encoder.layer.", "encoder.layers."),
    (".attention.attention.query.", ".attention.q_proj."),
    (".attention.attention.key.", ".attention.k_proj."),
    (".attention.attention.value.", ".attention.v_proj."),
    (".attention.output.dense.", ".attention.o_proj."),  # before .output.dense.
    (".intermediate.dense.", ".mlp.fc1."),
    (".output.dense.", ".mlp.fc2."),
]

def rename_encoder_keys(sd):
    if "encoder.encoder.layer.0.attention.attention.query.weight" in model.state_dict():
        return sd  # transformers<5, the checkpoint already matches the model
    out = {}
    for key, value in sd.items():
        if key.startswith("encoder."):
            for old, new in vit_renames:
                key = key.replace(old, new)
        out[key] = value
    return out

sd = torch.load(src / "weights.pt", map_location="cpu", weights_only=True)
model.load_state_dict(rename_encoder_keys(sd), strict=True)

swm.wm.utils.save_pretrained(model, run_name=run_name, config=OmegaConf.create(cfg))

# sanity check: the checkpoint round-trips through the eval.py loading path
reloaded = swm.wm.utils.load_pretrained(run_name).state_dict()
assert all(torch.equal(v, reloaded[k]) for k, v in model.state_dict().items())
print(f"✅ ready, run with: python eval.py --config-name=pusht.yaml policy={run_name}")
PY
```

Then:

```bash
./get_model.sh
python eval.py --config-name=pusht.yaml policy=pusht/lewm
```

The remap was verified numerically, not just by shape: the same input through the converted model under `transformers==5.15.0` and under `4.57.6` (no remap) gives bit-identical embeddings and predictions.

If you'd rather not deal with any of this, pinning `uv pip install 'transformers<5'` also makes the original snippet work.

## On your question 4

With `stable-worldmodel>=0.1.0` you can skip the conversion entirely *if* you're happy running their LeWM implementation instead of this repo's `jepa.py`:

```python
swm.wm.utils.load_pretrained("quentinll/lewm-pusht")  # fetches config.json + weights.pt from HF
```

Same `transformers` caveat applies, since it also ends up calling `load_state_dict` on those weights.

-----------------------------------------------------

Hope this helps you or the developers.

**Comment by flamma7** [comment]
It looks like there are still are versioning issues. Here are steps that worked for me in evaluating PushT with LeWM. Below I've attached a `requirements_working.txt` that uses `stable-worldmodel==0.1.1` (the latest release at this time), `stable-pretraining==0.1.8`, `datasets==5.0.1`, and `transformers==4.57.1` (there is a breaking change in v5). 

1. Install the working package recipe

[requirements_working.txt](https://github.com/user-attachments/files/31182150/requirements_working.txt)

```bash
uv venv --python=3.10 .venv
source .venv/bin/activate
uv pip install -r requirements_working.txt
```

2. Set your `STABLEWM_HOME` variable and create the required directories

```bash
export STABLEWM_HOME=/path/to/your/stablewm-home

mkdir -p "$STABLEWM_HOME/datasets"
mkdir -p "$STABLEWM_HOME/checkpoints/models--quentinll--lewm-pusht"
```

3. Pull the original trained LeWM model directly into SWM's expected checkpoint location. You can try `--revision main` but I'll include this commit hash to assure it works with the frozen Python packages.

```bash
hf download quentinll/lewm-pusht \
    config.json weights.pt \
    --revision 22b330c28c27ead4bfd1888615af1340e3fe9052 \
    --local-dir "$STABLEWM_HOME/checkpoints/models--quentinll--lewm-pusht"
```

4. Confirm your package versions can load LeWM.

[load_model.py](https://github.com/user-attachments/files/31182415/load_model.py)

```bash
python load_model.py
```

5. Now you can load LeWM with swm `0.1.1`. To download and prep the Push-T dataset directly in SWM's expected `datasets` folder, run the following:

```bash
hf download quentinll/lewm-pusht \
    pusht_expert_train.h5.zst \
    --repo-type dataset \
    --local-dir "$STABLEWM_HOME/datasets"

zstd -d \
    "$STABLEWM_HOME/datasets/pusht_expert_train.h5.zst" \
    -o "$STABLEWM_HOME/datasets/pusht_expert_train.h5"
```

6. To evaluate, checkout/clone version `0.1.1` of SWM

```bash
git clone --branch 0.1.1 --depth 1 \
    https://github.com/galilai-group/stable-worldmodel.git \
    stable-worldmodel-0.1.1
```

7. Run the evaluation

```bash
cd stable-worldmodel-0.1.1

python scripts/plan/eval_wm.py \
    policy=quentinll/lewm-pusht
```
